apis/database/v1alpha1/redshift_types.go

```diff
@@ -0,0 +1,713 @@
+	// contains the information the Amazon Redshift cluster can use to retrieve
+	// and store keys in an HSM.
+	// +optional
+	HSMConfigurationIdentifier *string `json:"hsmConfigurationIdentifier,omitempty"`
+
+	// IAMRoles is a list of AWS Identity and Access Management (IAM) roles that can be used
+	// by the cluster to access other AWS services. You must supply the IAM roles
+	// in their Amazon Resource Name (ARN) format. You can supply up to 10 IAM roles
+	// in a single request.
+	// A cluster can have up to 10 IAM roles associated with it at any time.
+	// kubebuilder:validation:MaxItems=10
+	// +optional
+	IAMRoles []string `json:"iamRoles,omitempty"`
+
+	// KMSKeyID for an encrypted DB cluster.
+	// The KMS key identifier is the Amazon Resource Name (ARN) for the KMS encryption
+	// key. If you are creating a DB cluster with the same AWS account that owns
+	// the KMS encryption key used to encrypt the new DB cluster, then you can
+	// use the KMS key alias instead of the ARN for the KM encryption key.
+	// +optional
+	KMSKeyID *string `json:"kmsKeyID,omitempty"`
+
+	// MaintenanceTrackName an optional parameter for the name of the maintenance track for the cluster.
+	// +optional
+	MaintenanceTrackName *string `json:"maintenanceTrackName,omitempty"`
+
+	// The default number of days to retain a manual snapshot. If the value is -1,
+	// the snapshot is retained indefinitely. This setting doesn't change the retention
+	// period of existing snapshots.
+	// default=1
+	// +kubebuilder:validation:Maximum=3653
+	// +optional
+	ManualSnapshotRetentionPeriod *int64 `json:"manualSnapshotRetentionPeriod,omitempty"`
+
+	// MasterUserPassword is the password associated with the master user account
+	// for the cluster that is being created.
+	// Constraints:
+	//    * Must be between 8 and 64 characters in length.
+	//    * Must contain at least one uppercase letter.
+	//    * Must contain at least one lowercase letter.
+	//    * Must contain one number.
+	//    * Can be any printable ASCII character (ASCII code 33 to 126) except '
+	//    (single quote), " (double quote), \, /, @, or space.
+	// +optional
+	NewMasterUserPassword *string `json:"masterUserPassword,omitempty"`
+
+	// MasterUsername is the user name associated with the master user account for the cluster that
+	// is being created.
+	// Constraints:
+	//    * Must be 1 - 128 alphanumeric characters. The user name can't be PUBLIC.
+	//    * First character must be a letter.
+	//    * Cannot be a reserved word. A list of reserved words can be found in
+	//    Reserved Words (https://docs.aws.amazon.com/redshift/latest/dg/r_pg_keywords.html)
+	//    in the Amazon Redshift Database Developer Guide.
+	// +immutable
+	MasterUsername *string `json:"masterUsername"`
+
+	// NewClusterIdentifier is the new identifier you want to use for the cluster.
+	// +optional
+	NewClusterIdentifier *string `json:"newClusterIdentifier,omitempty"`
```

Comment:
I think this should be the `external-name` annotation, so it should be omitted in body. https://github.com/crossplane/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#custom-resource-instance-name

Reply:
@hasheddan What you are referring to is `Cluster Identifier` and that is implemented through the `external-name` annotation. `NewClusterIdentifier` is for when the user wants to rename the cluster.

Reply:
Ah I see. This is a very imperative operation by the AWS API here. Typically we would want to make this more declarative by just allowing the user to change the field and then reconciling the external state to match. This could be done for something like `MasterUserPassword`, but since the identifier is used to actually get the external state before we modify it, we clearly cannot do the same. I wonder if we should just omit this functionality for now. Any thoughts @muvaf?